source/common/runtime/runtime_impl.cc

```diff
@@ -510,9 +513,9 @@ void LoaderImpl::loadNewSnapshot() {
     return std::static_pointer_cast<ThreadLocal::ThreadLocalObject>(ptr);
   });
 
-  {
-    absl::MutexLock lock(&snapshot_mutex_);
-    thread_safe_snapshot_ = ptr;
+  for (ThreadSafeSnapshot& shard : thread_safe_snapshots_) {
```

Comment:
Can you share some contention profiles that provide some insight about why this lock must be sharded?

Reply:
Profile as follows without the sharded lock. The worker threads are created outside of Envoy server for handling HTTP/2 requests,
![image](https://user-images.githubusercontent.com/16513687/107412418-a300fe00-6add-11eb-9b2b-c7add0a9fe44.png)

Reply:
> The worker threads are created outside of Envoy server

Could you elaborate please? As in for this profile?

Reply:
A bit description on the issue. Basically, we have workload that cannot be fulfilled by the Envoy server (thread-model, internal api, connection management, etc etc.). So what we are doing is in our binary is,
- create an Envoy server via MainCommonBase (instantiate tls and calls set on tls) (also initialize runtime singleton)
- then create a bunch of dispatcher threads (can no longer be register to tls as "set" has been called)
- then create Envoy HTTP/2 client connections on the dispatcher threads, to handle HTTP/2 requests
- such HTTP/2 connections will call Runtime to check features, for instance, for flooding prevention. Those calls leads to the slowpath in threadsafeSnapshot(), effectively hitting a single global lock
- with all the HTTP/2 connections hitting the global lock, the locking/unlocking consumes 32% of CPU.

Reply:
@yanavlasov 

ClientConnectionImpl should have latched that runtime flag at construction time.

skip_encoding_empty_trailers_ further up in the same file follows the right pattern.

Reply:
Could you see if patching in https://github.com/envoyproxy/envoy/pull/15005 helps reduce contention?

Reply:
It will probably improve the contention as well as reduce the cost of lookup itself, which is great. However it might not be sufficient  to address the contention, as there are a few other code paths. For instance there are heavy calls of TypedHeaderMapImpl::TypedHeaderMap::Impl==>Runtime::getInteger, as well as 
TimerImpl::TimerImpl==>Runtime::runtimeFeatureEnabled, both might be called for each request.

Reply:
I need to deprecate the use in TimerImpl.  That runtime feature was introduced long ago and has been verified to be safe.

I missed the TypedHeaderMapImpl case in the profile; I expect that code will trigger about twice per request.  Interesting.

Reply:
Yep, any feature check >=once per request will lead to regression to us. So it seems we should have both changes like #15005, as well as the current change, to reach maximum reliability? I'd be happy to rerun loadtests (without the current optimization) after these feature-check are fixed, to see if there is anything else missed.

Reply:
Note that the profile data is from an internal product that is using Envoy's guts in a way that was not intended. This is not a problem in Envoy. This is arguably not a good example of changing OSS sources.

Reply:
Curious whether dispatchers are used in other Envoy library applications in a similar way, e.g. for Envoy Mobile @alyssawilk @junr03 

The complexity of this optimization seems low and localized, so I'm not strongly opposed to it landing in OSS.

Reply:
My objection is that the threads in question don't have proper TLS support.  We're going to have further problems down the line and is going to become a maintenance tax.

I'm also curious about how Envoy mobile threading works and if they have similar problems.

Reply:
@yashengy did you have any luck in replacing `Runtime` with something that made sense for your use case? I.e. that doesn't have any concerns here?

Reply:
I didn't try to replace Runtime. Doing that will prevent us from creating Envoy server, because Runtime singleton can be initialized only once (check-fail otherwise). Discussed with antoniovicente@, and we may go with an internal patch on this change for now. Will see if this change is needed for other usages.

Reply:
What else do you want Envoy server for? In your use case it seems this is generally undesirable (for cross-thread reasons at least).

Reply:
We'll need to act as an HTTP/2 server. We are also interested in doing some thread juggling, a feature that Envoy doesn't support today, and not needed by others. We'll dig into that further but nothing solid yet. So we don't want a private runtime to prevent this effort.

Reply:
Re: Envoy Mobile threading. Currently Envoy Mobile just runs Envoy's main thread for all control and data plane work. Data plane I/O is dispatched from mobile threads to Envoy via Event::Dispatcher's posts, and back to mobile threads via native models (dispatch queues in iOS and Executor's in Android).

Mobile work loads (at least the ones we have put through Envoy Mobile up until now) are not highly concurrent, so we have avoided having multiple Envoy dispatch loops running.